test/widgets/profile_test.dart

```diff
@@ -0,0 +1,269 @@
+      check(testBinding.takeLaunchUrlCalls())
+        .single.equals((url: Uri.parse(testUrl), mode: expectedMode));
+    });
+
+    testWidgets('page builds; external link type navigates away', (WidgetTester tester) async {
+      final user = eg.user(userId: 1, profileData: {
+        0: ProfileFieldUserData(value: 'externalValue'),
+      });
+
+      await setupPage(tester,
+        users: [user],
+        pageUserId: user.userId,
+        customProfileFields: [mkCustomProfileField(0, CustomProfileFieldType.externalAccount, fieldData:'{"subtype": "external1"}')],
+        realmDefaultExternalAccounts: {
+          'external1': RealmDefaultExternalAccount(
+            name: 'external1',
+            text: '',
+            hint: '',
+            urlPattern: 'http://example/%(username)s')},
+      );
+
+      await tester.tap(find.text('externalValue'));
+      final expectedMode = defaultTargetPlatform == TargetPlatform.android ?
+        LaunchMode.externalApplication : LaunchMode.platformDefault;
+      check(testBinding.takeLaunchUrlCalls())
+        .single.equals((url: Uri.parse('http://example/externalValue'), mode: expectedMode));
+    });
+
+    testWidgets('page builds; user links to profile', (WidgetTester tester) async {
+      final users = [
+        eg.user(userId: 1, profileData: {
+          0: ProfileFieldUserData(value: '[2]'),
+        }),
+        eg.user(userId: 2, fullName: 'test user'),
+      ];
+      final pushedRoutes = <Route<dynamic>>[];
+      final testNavObserver = TestNavigatorObserver()
+        ..onPushed = (route, prevRoute) => pushedRoutes.add(route);
+
+      await setupPage(tester,
+        users: users,
+        pageUserId: 1,
+        customProfileFields: [mkCustomProfileField(0, CustomProfileFieldType.user)],
+        navigatorObserver: testNavObserver,
+      );
+
+      final textFinder = find.text('test user');
+      check(textFinder.evaluate()).length.equals(1);
+      final fieldContainer = find.ancestor(of: textFinder, matching: find.byType(Column)).first;
+      final targetWidget = find.descendant(of: fieldContainer, matching:find.byType(Avatar));
+      await tester.tap(targetWidget, warnIfMissed: false);
+      check(pushedRoutes).last.isA<WidgetRoute>().page.isA<ProfilePage>().userId.equals(2);
+    });
+
+    testWidgets('page builds; dm links to correct narrow', (WidgetTester tester) async {
+      final pushedRoutes = <Route<dynamic>>[];
+      final testNavObserver = TestNavigatorObserver()
+        ..onPushed = (route, prevRoute) => pushedRoutes.add(route);
+      final expectedNarrow = DmNarrow(
+        allRecipientIds: [eg.selfUser.userId, 1]..sort(),
+        selfUserId: eg.selfUser.userId);
+
+      await setupPage(tester,
+        users: [eg.user(userId: 1)],
+        pageUserId: 1,
+        navigatorObserver: testNavObserver,
+      );
+
+      final targetWidget = find.byIcon(Icons.email);
+      await tester.tap(targetWidget, warnIfMissed: false);
```

Comment:
Hmm, why `warnIfMissed: false`? The event does make it to the button, right?

Reply:
Great, that sounds like a good solution.


> I think depending on the test circumstances (the dimension of the virtual screen) the button might not be scrolled into view.

Hmm, I see. That shouldn't vary when running the tests, though — `flutter test` should always default to 800x600, and if any test case leaves it in a different state for the next test case then that's a bug in the test (and IIRC will also get flagged automatically). So the `tester.ensureVisible` may be a good step for ensuring the test is robust to unrelated changes in the code, but there shouldn't have been anything nondeterministic about the size.

Unless you're running the test in `flutter run` for debugging, that is. There's a known issue that `flutter run` on a test fails to preserve the test's screen size.